src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs

```diff
@@ -3821,6 +3820,300 @@ internal new NamedTypeSymbol CreateNativeIntegerTypeSymbol(bool signed)
+
+            var binaryOperatorName = OperatorFacts.BinaryOperatorNameFromSyntaxKindIfAny(syntaxKind, SyntaxFacts.IsCheckedOperator(name));
+            if (binaryOperatorName != name)
+                throw new ArgumentException(string.Format(CodeAnalysisResources._0_was_not_a_valid_built_in_operator_name, name), nameof(name));
+
+            // Lang specific checks to ensure this is an acceptable operator.
+            validateSignature();
+
+            return new SynthesizedIntrinsicOperatorSymbol(csharpLeftType, name, csharpRightType, csharpReturnType).GetPublicSymbol();
+
+            void validateSignature()
+            {
+                // Dynamic built-in operators allow virtually all operations with all types.  So we do no further checking here.
+                if (csharpReturnType.TypeKind is TypeKind.Dynamic ||
+                    csharpLeftType.TypeKind is TypeKind.Dynamic ||
+                    csharpReturnType.TypeKind is TypeKind.Dynamic)
+                {
+                    return;
+                }
+
+                // Use fast-path check to see if this types are ok.
+                var binaryKind = Binder.SyntaxKindToBinaryOperatorKind(SyntaxFacts.GetBinaryExpression(syntaxKind));
+
+                if (csharpReturnType.SpecialType != SpecialType.None &&
+                    csharpLeftType.SpecialType != SpecialType.None &&
+                    csharpRightType.SpecialType != SpecialType.None)
+                {
+                    var easyOutBinaryKind = OverloadResolution.BinopEasyOut.OpKind(binaryKind, csharpLeftType, csharpRightType);
+
+                    if (easyOutBinaryKind != BinaryOperatorKind.Error)
+                    {
+                        var signature = this.builtInOperators.GetSignature(easyOutBinaryKind);
+                        if (csharpReturnType.SpecialType == signature.ReturnType.SpecialType &&
+                            csharpLeftType.SpecialType == signature.LeftType.SpecialType &&
+                            csharpRightType.SpecialType == signature.RightType.SpecialType)
+                        {
+                            return;
+                        }
+                    }
+                }
+
+                // bool operator ==(object, object) is legal.
+                // bool operator !=(object, object) is legal.
+                // bool operator ==(Delegate, Delegate) is legal.
+                // bool operator !=(Delegate, Delegate) is legal.
+                if (binaryKind is BinaryOperatorKind.Equal or BinaryOperatorKind.NotEqual &&
+                    csharpReturnType.SpecialType is SpecialType.System_Boolean)
+                {
+                    if ((csharpLeftType.SpecialType is SpecialType.System_Object && csharpRightType.SpecialType is SpecialType.System_Object) ||
+                        (csharpLeftType.SpecialType is SpecialType.System_Delegate && csharpRightType.SpecialType is SpecialType.System_Delegate))
```

Comment:
Small suggestion: could make this a bit nicer with tuple patterns:

```cs
(csharpLeftType.SpecialType, csharpRightType.SpecialType) is (SpecialType.System_Object, SpecialType.System_Object)
                                                          or (SpecialType.System_Delegate, SpecialType.System_Delegate)
```

Reply:
nice!